examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp

```diff
@@ -172,35 +213,37 @@ EmberAfStatus OTAProviderExample::HandleQueryImage(chip::app::CommandHandler * c
 
     switch (mQueryImageBehavior)
     {
-    case kRespondWithUnknown:
+    case kRespondWithUnknown: {
         // This use-case is a subset of the ota-candidates-file option.
         // Can be removed once all other platforms start using the ota-candidates-file method.
         if (strlen(mOTAFilePath) > 0) // If OTA file is directly provided
         {
-            // TODO: Following details shall be read from the OTA file
-
-            // If software version is provided using command line then use it
-            if (mSoftwareVersion.HasValue())
-            {
-                newSoftwareVersion = mSoftwareVersion.Value();
-            }
-
-            // If software version string is provided using command line then use it
-            if (mSoftwareVersionString)
-            {
-                newSoftwareVersionString = mSoftwareVersionString;
-            }
+            // Parse the header and set version info based on the header
+            VerifyOrDie(ParseOTAHeader(mOTAFilePath, header) == true);
+            VerifyOrDie(SW_VER_STR_MAX_LEN > header.mSoftwareVersionString.size());
+            newSoftwareVersion = header.mSoftwareVersion;
+            strncpy(newSoftwareVersionString, header.mSoftwareVersionString.data(), header.mSoftwareVersionString.size());
 
             queryStatus = OTAQueryStatus::kUpdateAvailable;
         }
         else if (!mCandidates.empty()) // If list of OTA candidates is supplied instead
         {
             if (SelectOTACandidate(commandData.vendorId, commandData.productId, commandData.softwareVersion, candidate))
             {
-                newSoftwareVersion       = candidate.softwareVersion;
-                newSoftwareVersionString = candidate.softwareVersionString;
-                otaFilePath              = candidate.otaURL;
-                queryStatus              = OTAQueryStatus::kUpdateAvailable;
+                // Parse the header and verify version info matches
+                VerifyOrDie(ParseOTAHeader(candidate.otaURL, header) == true);
+                VerifyOrDie(candidate.softwareVersion == header.mSoftwareVersion);
+                VerifyOrDie(strlen(candidate.softwareVersionString) == header.mSoftwareVersionString.size());
+                VerifyOrDie(memcmp(candidate.softwareVersionString, header.mSoftwareVersionString.data(),
+                                   header.mSoftwareVersionString.size()) == 0);
+                VerifyOrDie(SW_VER_STR_MAX_LEN > strlen(candidate.softwareVersionString));
```

Comment:
```suggestion
                VerifyOrDie(sizeof(newSoftwareVersionString) > strlen(candidate.softwareVersionString));
```
Also, why the strlen bit here and not `header.mSoftwareVersionString.size()` which we already checked is the same but is faster?

Reply:
Actually, making a change to this PR such that we're not verifying here (but at an earlier step) so will no longer be parsing the header here and therefore using `strlen` after all